### PR TITLE
Change DrupalInstaller locations to use contrib folder

### PR DESCRIPTION
**Pull request by edysmp**
See https://github.com/composer/installers/issues/293

### CODE REVIEW

**Comment by jadb** [comment]
Please see: https://github.com/composer/installers/pull/285#issuecomment-200423753

**Comment by edysmp** [comment]
@jadb Fixed unit test :)

**Comment by bojanz** [comment]
This makes sense. +1

File: tests/Composer/Installers/Test/InstallerTest.php [hunk 1]
```diff
@@ -223,10 +223,15 @@ public function dataForTestInstallPath()
             array('dokuwiki-plugin', 'lib/plugins/someplugin/', 'author/someplugin'),
             array('dokuwiki-template', 'lib/tpl/sometemplate/', 'author/sometemplate'),
             array('dolibarr-module', 'htdocs/custom/my_module/', 'shama/my_module'),
-            array('drupal-module', 'modules/my_module/', 'shama/my_module'),
-            array('drupal-theme', 'themes/my_module/', 'shama/my_module'),
-            array('drupal-profile', 'profiles/my_module/', 'shama/my_module'),
             array('drupal-drush', 'drush/my_module/', 'shama/my_module'),
+            array('drupal-module', 'modules/contrib/my_module/', 'shama/my_module'),
+            array('drupal-custom-module', 'modules/custom/my_module/', 'shama/my_module'),
+            array('drupal-theme', 'themes/contrib/my_module/', 'shama/my_module'),
```
Comment (cweagans) [inline tests/Composer/Installers/Test/InstallerTest.php]: `shama/my_theme`, perhaps?

File: tests/Composer/Installers/Test/InstallerTest.php [hunk 2]
```diff
@@ -223,10 +223,15 @@ public function dataForTestInstallPath()
...
+            array('drupal-custom-theme', 'themes/custom/my_module/', 'shama/my_module'),
+            array('drupal-profile', 'profiles/contrib/my_module/', 'shama/my_module'),
```
Comment (cweagans) [inline tests/Composer/Installers/Test/InstallerTest.php]: `shama/my_profile`, perhaps?

File: tests/Composer/Installers/Test/InstallerTest.php [hunk 3]
```diff
@@ -223,10 +223,15 @@ public function dataForTestInstallPath()
...
+            array('drupal-custom-profile', 'profiles/custom/my_module/', 'shama/my_module'),
+            array('drupal-theme-engine', 'themes/engines/contrib/my_module/', 'shama/my_module'),
```
Comment (cweagans) [inline tests/Composer/Installers/Test/InstallerTest.php]: `shama/my_theme_engine`, perhaps?

**Comment by cweagans** [comment]
Sorry for the noise. Misunderstood what was happening in the test file. :+1: from me. This looks great.

File: src/Composer/Installers/DrupalInstaller.php [hunk 4]
```diff
@@ -3,14 +3,17 @@
...
-        'module'    => 'modules/{$name}/',
-        'theme'     => 'themes/{$name}/',
-        'library'   => 'libraries/{$name}/',
-        'profile'   => 'profiles/{$name}/',
-        'drush'     => 'drush/{$name}/',
-	    'custom-theme' => 'themes/custom/{$name}/',
-	    'custom-module' => 'modules/custom/{$name}',
-    );
+	protected $locations = array(
+		'core'                => 'core/',
```
Comment (slbmeh) [inline src/Composer/Installers/DrupalInstaller.php]: `drupal-core`, `drupal-module`, and `drupal-theme` were bold in the readme... Which it states they are stable and install paths will not be changed.

**Comment by typhonius** [comment]
👍 from me for this patch.

**Comment by rujiali** [comment]
Looks good to me.

**Comment by swim** [comment]
At time of writing the current composer file shipped with D8 core (8.6.3) defines the following installer paths.
`
"installer-paths": {
  "core": ["type:drupal-core"],
  "modules/contrib/{$name}": ["type:drupal-module"],
  "profiles/contrib/{$name}": ["type:drupal-profile"],
  "themes/contrib/{$name}": ["type:drupal-theme"],
  "drush/contrib/{$name}": ["type:drupal-drush"],
  "modules/custom/{$name}": ["type:drupal-custom-module"],
  "themes/custom/{$name}": ["type:drupal-custom-theme"]
}
`
The PR seems more aligned to the above 👍

**Comment by Seldaek** [comment]
Going to close this as I assume this is not needed anymore. If it is then it should be discussed in concert with the drupal team, as at this point so many Drupal sites are out there depending on the behavior here that I am not really comfortable changing the behavior without consulting people who are more in the know.

**Pull request closed without merging**